### PR TITLE
[4.0] Fix the vendor css for the template

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

Joomla has a very well established pattern for overridding CSS and JS assets. You place the override file in the template specific folder and magic happens.

For some reason, instead of promoting modularity and the core functionality we end up making the css of the template a huge garbage bin, where anyone can append some css.

This should stop as the produced css is not maintainable and is spaggetti code that nobody could figure out where could be the source of a bug (cascading is already hard let's not make it even harder). Not to mention that performance also has a negative impact but that's minor in this case...

### Summary of Changes
All the files that are loaded through an HTMLHelper should NOT be overridden with code internally to the template. This PR reverts it for the following external resources (vendor folder):
- awesomplete
- choicesjs
- font awesome
- minicolors
- joomla-alert
- joomla-tab

### Testing Instructions
Check if anything is broken in the backend 
(it shouldn't be as I'm still importing the original css, I know but then this PR is becoming a major refactor...)

### Documentation Changes Required

PS there should be some QA for the scss of the template, the code is nowhere near beta state, just saying

### CODE REVIEW

File: administrator/templates/atum/scss/template.scss [hunk 1]
```diff
@@ -15,10 +15,7 @@
 @import "fonts/fontawesome";
 
 // Alert
-@import "../../../../build/media_source/system/scss/jalert";
-
-// jQuery Minicolors
-@import "../../../../build/media_source/system/scss/jquery-minicolors";
+//@import "../../../../build/media_source/system/scss/jalert"; NOT USED...
```
Comment (zero-24) [inline administrator/templates/atum/scss/template.scss]: when not used we can remove it right?

Comment (dgrammatiko) [reply administrator/templates/atum/scss/template.scss]: There was a reason that this file was introduced in the first place so calling @C-Lodder and @ciar4n

Comment (C-Lodder) [reply administrator/templates/atum/scss/template.scss]: Can't remember why I added this as it was 3 years ago. But if it's not being used then remove it

Comment (dgrammatiko) [reply administrator/templates/atum/scss/template.scss]: @C-Lodder if you can confirm that this can be removed I will do so

Comment (C-Lodder) [reply administrator/templates/atum/scss/template.scss]: @dgrammatiko You can remove it completely. Just look properly and I added it to support the `.alert-message` being generated but think Joomla have fixed all their alert classes now.

**Comment by ghost** [comment]
Please resolve conflicting files so this PR can get tested at Worldwide Pizza, Bugs & Fun, October 19th

File: administrator/templates/atum/scss/template.scss (hunk 1 above)
Comment (dgrammatiko) [reply administrator/templates/atum/scss/template.scss]: done

**Comment by Quy** [comment]
FontAwesome icons don't appear.

**Comment by dgrammatiko** [comment]
@Quy do you have the css file `administrator/templates/atum/css/vendor/fontawesome-free/fontawesome.min.css` ? If so do you also get the actual font?

**Comment by Quy** [comment]
Installed your branch. There is no `vendor` directory.

**Comment by C-Lodder** [comment]
@Quy Did you run `npm i`?

**Comment by Quy** [comment]
Yes, both `npm i` and `npm ci`. Still no go.

**Comment by dgrammatiko** [comment]
@Quy what's your OS?

**Comment by Quy** [comment]
Testing locally with XAMPP on Windows 10 Home.

**Comment by dgrammatiko** [comment]
@Quy thanks, I'll check why the build tools are failing to generate the css for winDOS

**Comment by dgrammatiko** [comment]
@Quy can you give it another try?

File: administrator/templates/atum/scss/vendor/fontawesome-free/fontawesome.scss [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+@import '../../../../../../node_modules/@fortawesome/fontawesome-free/scss/variables';
+@import '../../../../../../node_modules/@fortawesome/fontawesome-free/scss/mixins';
+
+@font-face {
+  font-family: 'Font Awesome 5 Free';
+  font-style: normal;
+  font-weight: 400;
+  font-display: $fa-font-display;
+  src: url('../../../../../../../media/vendor/fontawesome-free/webfonts/fa-regular-400.woff2') format('woff2'),
```
Comment (Quy) [inline administrator/templates/atum/scss/vendor/fontawesome-free/fontawesome.scss]: I had to reduce by one level to make the icons appear.

**Comment by Quy** [comment]
Almost there. Here are a few more that don't appear.

26077a
26077b

**Comment by dgrammatiko** [comment]
@Quy thank you, should fine now

**Comment by Quy** [comment]
They show up but don't display properly.

26077c

**Comment by Quy** [comment]
It is still an issue. Please let me know when to test again. Thanks.

**Comment by dgrammatiko** [comment]
Did you run `npm I` and cleaned the browser cache? Works fine here...

**Comment by Quy** [comment]
Yes. Some display correctly and others don't.

26077d

**Comment by dgrammatiko** [comment]
what's your browser?

**Comment by Quy** [comment]
Tested with Firefox 69.0.3 and Chrome 77.0.3865.120. Icons with class `icon-` don't display properly.

**Comment by ciar4n** [comment]
@dgrammatiko 

FontAwesome 5 now requires a font-weight of 900 on the pseudo class. Try adding the following to the start of the _icons.scss ...

```
[class*=" icon-"]::before,
[class^="icon-"]::before,
.btn::after {
  font-weight: 900;
}
```

**Comment by Quy** [comment]
@ciar4n It did the trick. A few more then we should be good.

26077login

26077help

**Comment by infograf768** [comment]
hmm
I can't get npm ci to work here
```
Internal Error: File to read not found or unreadable: /Applications/MAMP/htdocs/newfolder/joomla40/administrator/templates/atum/scss/vendor/minicolors/minicolors.scss

sh: line 1: 25356 Segmentation fault: 11  node build.js --compile-css
```
If I redo, then I get
```
Internal Error: File to read not found or unreadable: /Applications/MAMP/htdocs/newfolder/joomla40/administrator/templates/atum/scss/vendor/choicesjs/choicesjs.scss

sh: line 1: 25484 Segmentation fault: 11  node build.js --compile-css
```

**Comment by nadjak77** [comment]
After applying patch and run npm i some icons are missing:
image
image

**Comment by nadjak77** [comment]
I have tested this item :red_circle: unsuccessfully on 0afab1713282fb9d090ec462807bcd4d44945620<br /><br />apply patch
run npm i
some icons are missing<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by dgrammatiko** [comment]
@nadjak77 can you create a gist with the font awesome css?

**Comment by nadjak77** [comment]
I hope this is the right file you want to see:
https://gist.github.com/nadjak77/cc199b47dc8309c16341a3776c817048
this is the template.css with your patch

**Comment by dgrammatiko** [comment]
@nadjak77 thanks, but I was referring to the actually css for font awesome. Anyways I think I figure out what was missing here

File: administrator/templates/atum/scss/vendor/fontawesome-free/fontawesome.scss [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Override the font path
+$fa-font-path: "../../../../../../../../media/vendor/fontawesome-free/webfonts" !default;
```
Comment (Quy) [inline administrator/templates/atum/scss/vendor/fontawesome-free/fontawesome.scss]:
```suggestion
$fa-font-path: "../../../../../../media/vendor/fontawesome-free/webfonts" !default;
```
It is now working with this change.

Comment (dgrammatiko) [reply administrator/templates/atum/scss/vendor/fontawesome-free/fontawesome.scss]: Thanks

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on bfd6ffe67fa38d8b050fd089fd90b7f3a85836ce<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on bfd6ffe67fa38d8b050fd089fd90b7f3a85836ce<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by SharkyKZ** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by SharkyKZ** [comment]
I have tested this item :red_circle: unsuccessfully on df4929f85efb60f7803dc36650146b55d2d631ef<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by SharkyKZ** [comment]
Some `choicesjs` field styling is missing:
Untitled

**Comment by dgrammatiko** [comment]
@SharkyKZ bandaid applied.
@wilsonge please check the assets code, they seem to ignore the overrides. Was that intentional?

**Comment by SharkyKZ** [comment]
Styling is still a little broken
Screenshot_2019-10-22 Articles Edit - Joomla - Administration

**Comment by SharkyKZ** [comment]
About asset issues you'll have to ask @Fedik.

**Comment by Fedik** [comment]
> @wilsonge please check the assets code, they seem to ignore the overrides. Was that intentional?

need more details, what exactly problem,
what you do and what you get?

**Comment by dgrammatiko** [comment]
> need more details, what exactly problem
what you do and what you get?

@Fedik should be easy to replicate: 
using the 4.0-Dev branch, In the atum/css folder create create a folder vendor/fontawesome-free. In this newly created folder create a css file named font awesome.css with some dummy content. The expected behavior is that the newly created css should be delivered instead of the media/vendor/fontawesome-free/...
But it isn’t

**Comment by Fedik** [comment]
All libs from `/media/vendor` has strict path in its own asset, see `/media/vendor/joomla.asset.json`
```
{
      "package": "@fortawesome/fontawesome-free",
      "name": "fontawesome-free",
      "version": "5.9.0",
      "css": [
        "media/vendor/fontawesome-free/css/fontawesome.min.css"
    ]
},
```
(btw it should be "name": "fontawesome",)

It can be overridden by overriding the asset.
that why you had:
```
{
      "name": "fontawesome",
      "css": [
        "fontawesome.min.css"
      ]
}
```
in the template. This override asset  "fontawesome" to use another file(s).

**Comment by SharkyKZ** [comment]
There are still two instances of `choices.css` loaded.

**Comment by dgrammatiko** [comment]
> There are still two instances of choices.css loaded.

I know! The problem is that the original is loaded with webAssets, that for some reason doesn't pick up the override. If I try to override the css then the JS ends up not loading. I think @Fedik already has a PR that gives more granular access but till that PR is merged the original (from media) css file is also loaded here.

**Comment by Fedik** [comment]
well, it sounds like it better to allow to override vendor files in "old way" also, for j4
I try to update that PR some time later, or I will do another one, will see

**Comment by dgrammatiko** [comment]
@wilsonge is there any interest on this one or should I move on?

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on ab05717405a4a2b1728a26e3bfad0385bbced5e3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on ab05717405a4a2b1728a26e3bfad0385bbced5e3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26077">issues.joomla.org/tracker/joomla-cms/26077</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Comment by infograf768** [comment]
@dgrammatiko 
Looks like atum specific choice.css is not loaded in com_associations.
Screen Shot 2019-12-01 at 17 02 26

**Comment by dgrammatiko** [comment]
@infograf768 are the tags ok in that page?

**Comment by infograf768** [comment]
Same.

Screen Shot 2019-12-01 at 17 44 54

**Comment by dgrammatiko** [comment]
@infograf768 in the `atum/component.php`
add
after line 17
```php
$wa    = $this->getWebAssetManager();
``` 
and after line 30
```php

// Enable assets
$wa->enableAsset('template.atum.' . ($this->direction === 'rtl' ? 'rtl' : 'ltr'));
$wa->enableAsset('fontawesome-free');

// TODO: remove the following line whenever the assets are fixed to respect the ovverides
HTMLHelper::_('stylesheet', 'vendor/choicesjs/choices.css', array('version' => 'auto', 'relative' => true));
```

On a more serious note the assets for the iframes need some attention (eg the component.php needs to reflect the index.php)

**Comment by dgrammatiko** [comment]
@infograf768 check https://github.com/joomla/joomla-cms/pull/27187

**Comment by Fedik** [comment]
> $wa->enableAsset('fontawesome-free');

all you need:
```javascript
$wa->enableAsset('fontawesome-free')->enableAsset('choices');
```

**Comment by dgrammatiko** [comment]
@Fedik this didn't work as it expects me to override both the css and the javascript. I wanted only the css from the choices to be overriden

**Pull request merged**